plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/callables/FullySpecialKotlinFunctionBuilder.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2010-2024 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.embeddings.callables
+
+import org.jetbrains.kotlin.formver.conversion.StmtConversionContext
+import org.jetbrains.kotlin.formver.embeddings.expression.ExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.types.FunctionTypeEmbedding
+import org.jetbrains.kotlin.formver.viper.MangledName
+
+class FullySpecialKotlinFunctionBuilder {
+    private val byName = mutableMapOf<MangledName, FunctionEmbedding>()
+
+    fun withCallableType(
+        callableType: FunctionTypeEmbedding,
+        functionsBlock: SpecialKotlinFunctionBuilderWithCallableType.() -> Unit,
+    ) {
+        val builderWithCallableType = SpecialKotlinFunctionBuilderWithCallableType(callableType)
+
+        builderWithCallableType.functionsBlock()
+    }
+
+    fun addFunction(
+        callableType: FunctionTypeEmbedding,
+        vararg packageName: String,
+        className: String? = null,
+        name: String,
+        body: (List<ExpEmbedding>, StmtConversionContext) -> ExpEmbedding,
+    ) {
+        withCallableType(callableType) {
+            addFunction(*packageName, className = className, name = name, body = body)
+        }
+    }
+
+    inner class SpecialKotlinFunctionBuilderWithCallableType(private val callableType: FunctionTypeEmbedding) {
+        fun addFunction(
+            vararg packageName: String,
+            className: String? = null,
+            name: String,
+            body: (List<ExpEmbedding>, StmtConversionContext) -> ExpEmbedding,
+        ) {
+
+            val newFunction = object : FullySpecialKotlinFunction {
```

Comment:
At this point, should `FullySpecialKotlinFunction` not have an implementation which this thing passes all the data to, rather than creating a new object for each case?  I'm not sure what the trade-offs are.  (Maybe `FullySpecialKotlinFunction` should just be a class?)